File: 120Triangle.cpp
// 动态规划思路
// 从下向上填表

class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int size = triangle.size();
        if(size <= 0){
            return 0;
        }
        if(size == 1){
            return triangle[0][0];
        }
        int table[size][size] = {0};
        for(int i = 0; i < size; i++){
            table[size-1][i] = triangle[size-1][i];
        }
        for(int j = size-2; j >= 0; j--){
            for(int k = 0; k < triangle[j].size(); k++){
                table[j][k] = min(table[j+1][k], table[j+1][k+1]) + triangle[j][k];
            }
        }
        return table[0][0];
    }
};

File: 120Triangle.py
class Solution:
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        if len(triangle) is 0:
            return 0
        res = [[0]*len(t) for t in triangle]
        tlen = len(triangle)
        for i in range(len(triangle)):
            for j in range(len(triangle[i])):
                if i is 0 and j is 0:
                    res[i][j] = triangle[i][j]
                elif j is 0:
                    res[i][j] = res[i-1][j] + triangle[i][j]
                elif j is len(triangle[i])-1:
                    res[i][j] = res[i-1][j-1] + triangle[i][j]
                else:
                    res[i][j] = min(res[i-1][j-1], res[i-1][j]) +  triangle[i][j]
        minret = res[tlen-1][0]
        for it in res[tlen-1]:
            if it < minret:
                minret = it
        return minret
